### ISSUE TITLE
T265 VIO coordinate system 

### ISSUE THREAD

**Issue by wedonsjtu** [Auterion/VIO#16]
Hi, I am faced with problems tuning t265 and px4. We install t265 camera towards the back of F450 airframe and 45 degress facing downward(while px4 is towards forward). like the picture shows. we change the bridge.launch file and set the extrinsic angle as "-0.11, 0, -0.06, 3.14, 0.785, 0. ""However we subscribe /mavros/local_position/pose and find out that the coordinate shows that x is still facing backward of the drone(indicate that the extrinsic parameter doesnt work). 
![589486181a7819a3da2249b48b18b66](https://user-images.githubusercontent.com/40729335/96338036-fd54f980-10bd-11eb-8ece-27f98224b799.jpg)
could u give a hint?

**Comment by hipforth** [comment]
It also doesn't work for me, maybe you can directly modify the source coordinate system
base_realsense_node.cpp

**Comment by DiegoHerrera1890** [comment]
Hello! May I ask you if you solved your problem? I am facing an issue related to T265 with PX4 and QGC but no one can help me. Can you explain to me how did you fly your drone indoors?

**Comment by bigbellmercy** [comment]
Hi~ It didn't work well too me. But, I succeeded it by the following environment:
 - PX4 firmware: v1.10.2
 - VIO commit: fb00d (by `git checkout fb00d` command in the VIO folder and then build again VIO)
  (This commit output `/mavros/odometry/out` instead of `/mavros/vision_pose/pose`)
 - EKF2_AID_MASK: 280
 - EKF2_EV_DELAY: 10 (10~35)

More info: In that environment, if the master commit (7484c) is used, `/mavros/vision_pose/pose` is output instead of `/mavros/odometry/out`, but its position estimate response is bad and weird.

Good luck, bros!

**Comment by nabatta** [comment]
@bigbellmercy @DiegoHerrera1890 Did you have issues where the VIO node always tells the Pixhawk it is facing due east?

**Comment by bigbellmercy** [comment]
@nabatta No, I didn't. How about check the console log for some error or warning.

**Comment by nabatta** [comment]
> @nabatta No, I didn't. How about check the console log for some error or warning?

I haven't had any warnings that would indicate that. When your code, does it change the heading of the Pixhawk at all? Thanks!

**Comment by bigbellmercy** [comment]
@nabatta Yes, the heading of PX4 changes as T265 changes.

**Comment by nabatta** [comment]
@bigbellmercy Would you be able to send me the launch file, "rs_t265.launch"? I cannot figure out what I have messed up. Thank you!

**Comment by DiegoHerrera1890** [comment]
I am facing the same... I hope to find a solution

**Comment by bigbellmercy** [comment]
@nabatta  and @DiegoHerrera1890, Yes, T265 don't know the absolute heading. But it knows relative change of heading from its beginning. But PX4 knows the absolute heading because Pixhawk has a magnetometer inside of it. So, the fused result of heading for Pixhawk and T265 both knows the absolute heading. Therefore, when the system is started, the heading displayed in QGroundControl is the actual absolute heading of the Pixhawk from the beginning. And the attitude and heading is more accurate than that with Pixhawk alone because T265 enhances the accuracy.

**Comment by nabatta** [comment]
@bigbellmercy This is how I thought it would work but it is not doing that for me. Instead, the VIO node overrides the Pixhawk and says that it is facing due East no matter where the drone is facing when I start the VIO node. I thought that it would take the Pixhawk heading and use that as the start point but I have not had any luck with that.

**Comment by 987428377** [comment]
Hi, master!
I try to reproduce this method to achieve t265 indoor positioning.
https://docs.px4.io/master/en/computer_version/visual_inertial_odometry.html
but I have installed librealsense V2. 43.0, and I input "dpkg -l | grep "realsense" | cut -d " " -f 3 | xargs sudo dpkg --purge" to uninstall it. And then I installed V2.42.0. When I input realsense-viewer, The window display version 2.4.2.
But an error occurred:
Built with LibRealSense v2.42.0
[ INFO] [1618483684.377621028]: Running with LibRealSense v2.43.0
[ WARN] [1618483684.378017539]: ***************************************************
[ WARN] [1618483684.378393063]: ** running with a different librealsense version **
[ WARN] [1618483684.378635271]: ** than the one the wrapper was compiled with! **
[ WARN] [1618483684.378998192]: ***************************************************
[ WARN] [1618483684.424316041]: No RealSense devices were found!

How do you configure it ? Looking forword to you guidance!

**Comment by bigbellmercy** [comment]
Hi @nabatta and @DiegoHerrera1890, this is my quite late reply for I found a solution yesterday. I think I now understood your problem for the East direction issue. And my previous explanation about heading seems to have some errors.

Here is one solution:
1) PX4 v1.10.2
2) PX4 VIO commit 434a069 (Date: 2021-10-28)
3) T265 toward down direction (the USB connector of T265 should be right side of the Pixhawk which is in forward direction)
4) Before turning on the T265, T265 should be placed toward East direction. After that turn on the T264. (to resolve the East direction issue)

In the condition above, it worked well as following:
1) The heading value matched well to real direction.
2) When I rotate the heading of the drone, its heading value changes well without any strange behavior.
3) In the QGroundControl Mavlink Inspector, the local position x, y and z values are correctly changed independently when the drone is moved in only one direction among NED (North(x)-East(y)-Down(z)) directions.
4) And the shape of the position graphs are clean without any overshoot and delay.

If you are flying indoor, EKF2_MAG_TYPE may be better to be changed from Automatic to None. And the EKF2_EV_DELAY was changed to 35 ms according to the measurement method in the site.

**Comment by DiegoHerrera1890** [comment]
@bigbellmercy 
Thanks for sharing with us this info.
According to the [PX4_VIO_guide](https://docs.px4.io/v1.11/en/computer_vision/visual_inertial_odometry.html) we should rotate the drone until we obtain a unit quaternion (w=1, x=y=z=0).

> Yaw the vehicle until the quaternion of the ODOMETRY message is very close to a unit quaternion (w=1, x=y=z=0).

    At this point the body frame is aligned with the reference frame of the external pose system.
    If you do not manage to get a quaternion close to the unit quaternion without rolling or pitching your vehicle, your frame probably still has a pitch or roll offset. Do not proceed if this is the case and check your coordinate frames again.

**Comment by MagicShawn** [comment]
I am facing the problem,the body coordinate has some strange ,when starting the offboard mode the plane body will auto change 90deg
